scripts/ASTImportTest.sh

```diff
@@ -1,105 +1,192 @@
 #!/usr/bin/env bash
+# ------------------------------------------------------------------------------
+# vim:ts=4:et
+# This file is part of solidity.
+#
+# solidity is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# solidity is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) solidity contributors.
+# ------------------------------------------------------------------------------
+# Bash script to test the import/exports.
+# ast import/export tests:
+#   - first exporting a .sol file to JSON, then loading it into the compiler
+#     and exporting it again. The second JSON should be identical to the first.
+
+set -euo pipefail
 
-set -e
-
-# Bash script to test the ast-import option of the compiler by
-# first exporting a .sol file to JSON, then loading it into the compiler
-# and exporting it again. The second JSON should be identical to the first
 READLINK=readlink
 if [[ "$OSTYPE" == "darwin"* ]]; then
     READLINK=greadlink
 fi
 REPO_ROOT=$(${READLINK} -f "$(dirname "$0")"/..)
 SOLIDITY_BUILD_DIR=${SOLIDITY_BUILD_DIR:-${REPO_ROOT}/build}
-SOLC=${SOLIDITY_BUILD_DIR}/solc/solc
-SPLITSOURCES=${REPO_ROOT}/scripts/splitSources.py
+SOLC="${SOLIDITY_BUILD_DIR}/solc/solc"
+SPLITSOURCES="${REPO_ROOT}/scripts/splitSources.py"
+
+# shellcheck source=scripts/common.sh
+source "${REPO_ROOT}/scripts/common.sh"
+
+function print_usage
+{
+    fail "Usage: ${0} ast [--exit-on-error]."
+}
+
+function print_used_commands
+{
+    local test_directory="$1"
+    local export_command="$2"
+    local import_command="$3"
+    printError "You can find the files used for this test here: ${test_directory}"
+    printError "Used commands for test:"
+    printError "# export"
+    echo "$ ${export_command}" >&2
+    printError "# import"
+    echo "$ ${import_command}" >&2
+}
+
+function print_stderr_stdout
+{
+    local error_message="$1"
+    local stderr_file="$2"
+    local stdout_file="$3"
+    printError "$error_message"
+    printError ""
+    printError "stderr:"
+    cat "$stderr_file" >&2
+    printError ""
+    printError "stdout:"
+    cat "$stdout_file" >&2
+}
+
+IMPORT_TEST_TYPE=
+EXIT_ON_ERROR=0
+for PARAM in "$@"
+do
+    case "$PARAM" in
+        ast) IMPORT_TEST_TYPE="ast" ;;
+        --exit-on-error) EXIT_ON_ERROR=1 ;;
+        *) print_usage ;;
+    esac
+done
 
 SYNTAXTESTS_DIR="${REPO_ROOT}/test/libsolidity/syntaxTests"
```

Comment:
In the original `ASTImportTest.sh` uses the syntax tests to verify the correct reimport. These contain a lot of not compilable contracts, so the question is, do we really need to use the syntax tests here? From my point of view it just increases the execution time and does not add any additional value. @Marenz You added this file, right? What is your opinion about this?

Reply:
Hmm.. I thought that we could just use the semantic tests here, but turns out that the reimport is failing then. 

```
ERROR: AST reimport failed (import) for input file /Users/alex/git/solidity/test/libsolidity/semanticTests/inlineAssembly/inline_assembly_function_call_assignment.sol.

stderr:
Failed to import AST: Expected "nodeType" to be of type string!

stdout:
You can find the files used for this test here: /var/folders/tv/qrf28_dn00l2ltkqk5q65gzc0000gn/T/tmp.6kqSMP4O
Used commands for test:
# export
$ /Users/alex/git/solidity/build/solc/solc --combined-json ast --pretty-json --json-indent 4 /Users/alex/git/solidity/test/libsolidity/semanticTests/inlineAssembly/inline_assembly_function_call_assignment.sol > expected.json
# import
$ /Users/alex/git/solidity/build/solc/solc --import-ast --combined-json ast --pretty-json --json-indent 4 expected.json
```

I would have expected that the reimport is also working on the semantic tests correctly.

Reply:
> I would have expected that the reimport is also working on the semantic tests correctly.

Looks like a bug in the AST import. I checked the output and all `nodeType` fields seem to actually be strings so I suspect that the problem is that we're expecting that field in some places but it's missing and we assume `null` by default - which is not a string.

Reply:
I reported this bug separately: #13684. We can add semantic tests later, when we deal with that bug.